lib/vault/api/auth.rb

```diff
@@ -186,6 +190,59 @@ def aws_ec2(role, pkcs7, nonce = nil)
       return secret
     end
 
+    # Authenticate via IAM EC2 method by providing a AWS CredentialProvider (either ECS, AssumeRole, etc.)
+    # If authentication is successful, the resulting token will be stored on the client and used
+    # for future requests.
+    #
+    # @example
+    #   Vault.auth.aws_ecs_iam("dev-role-iam", "us-east-1", Aws::AssumeRoleCredentials.new, "vault.example.com") #=> #<Vault::Secret lease_id="">
+    #
+    # @param [String] role
+    # @param [String] region
+    # @param [CredentialProvider] credentials_provider
+    #   https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/CredentialProvider.html
+    # @param [String] iam_auth_header_value optional
+    #
+    # @return [Secret]
+    def aws_iam(role, region, credentials_provider, iam_auth_header_value = IAM_SERVER_ID_HEADER)
+      require "aws-sigv4"
+      require "base64"
+
+      request_body   = 'Action=GetCallerIdentity&Version=2011-06-15'
+      request_url    = 'https://sts.amazonaws.com/'
+      request_method = 'POST'
+
+      vault_headers = {
+        'User-Agent' => Vault::Client::USER_AGENT,
+        'Content-Type' => 'application/x-www-form-urlencoded; charset=utf-8',
+        'X-Vault-AWSIAM-Server-Id' => iam_auth_header_value
```

Comment:
It's actually [X-Vault-AWS-IAM-Server-ID](https://github.com/hashicorp/vault/blob/e9bbdd9c2c5aea2605c06cbf52be41aacc80d12b/builtin/credential/aws/path_login.go#L1620) (hyphen between AWS and IAM)